File: 61.rotate-list.py
#
# @lc app=leetcode id=61 lang=python3
#
# [61] Rotate List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head,k):
        if not head : 
            return head
    # Get length
        length , tail =1 , head # we have make length to 1 , 
        # it will be one head and tail will be at same position
        while tail.next: # Untill Null
            tail=tail.next # we will move head pointer untill it reches null
            length+=1 # As it coves we increase length so we can count length
        k = k%length # For finding it's k we will use mod 

        if k==0:
            return head
        curr=head #we will assign curr to head value
        for i in range(length-k-1): # we will go till the index of that 
            curr=curr.next # increament head value 
        newHead=curr.next # storing it 
        curr.next=None #then once it will reach to null
        tail.next=head # we will swap as it's rotaion and make tail value to head
        return newHead
        
# @lc code=end

File: 61.rotate-list.rb
#
# @lc app=leetcode id=61 lang=ruby
#
# [61] Rotate List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} k
# @return {ListNode}
# def rotate_right(head, k):
#     class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

def rotateRight(self , head, k):
    # Calculate the length of the linked list
    length = 0
    current = head
    while current:
        length += 1
        current = current.next
    
    # Handle cases where the list is empty or has only one node
    if length == 0 or k % length == 0:
        return head
    
    # Find the new head position after rotation
    k %= length
    steps_to_new_head = length - k
    new_head = head
    for _ in range(steps_to_new_head - 1):
        new_head = new_head.next
    
    # Update pointers to achieve the rotation
    new_tail = new_head
    while new_tail.next:
        new_tail = new_tail.next
    new_tail.next = head
    head = new_head.next
    new_head.next = None
    
    return head

    

# @lc code=end

